pkg/syfs/syfs.go

```diff
@@ -0,0 +1,64 @@
+// Pakcage syfs provides functions to access singularity's file system
+// layout
+
+package syfs
+
+import (
+	"os"
+	"os/user"
+	"path/filepath"
+	"sync"
+
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+)
+
+const singularityDir = ".singularity"
+
+// cache contains the information for the current user
+var cache struct {
+	sync.Once
+	configDir string // singularity user configuration directory
+}
+
+// ConfigDir returns the directory where the singularity user
+// configuration and data is located.
+func ConfigDir() string {
+	cache.Do(func() {
+		cache.configDir = configDir()
+		sylog.Debugf("Using singularity directory %q", cache.configDir)
+	})
+
+	return cache.configDir
+}
+
+func configDir() string {
```

Comment:
I agree that we should return an error if we cannot lookup the current user. I don't like the idea of potentially creating user config directories at different locations based on `cwd` if we cannot resolve the user. I think that's enough reason to return an error and bail on execution.

Reply:
Thanks for looking at this.

The reason why configDir does not return an error is because it covers all the possible error conditions and provides fallbacks for it. I don't agree with what this is doing, but it's been ported over from the sypgp package. In particular, the choice of falling back to a .singularity directory in the current directory if the current user information is not available seems strange to me. The choice to return a relative .singularity path if the current directory cannot be obtained seems even stranger.

ConfigDirForUsername behaves differently because it's possible that the username in the argument does not exist, so it makes sense to return an error in that case. The _current_ user _should_ exist, and to be consistent with the behavior of ConfigDir, no error is returned in that case ever, it simply falls back to whatever ConfigDir returns. It would be strange to have ConfigDirForUsername(current_user) return something different than ConfigDir().